#  Analysis Report On:
## How Much of the World Has Access to the Internet?

This report was curated to aid the firm's principal with her presentation on the state of internet access in the world..

## 💾 The data

#### Tables compiled by the research team ([source](https://ourworldindata.org/internet)):

#### internet
- "Entity" - The name of the country, region, or group.
- "Code" - Unique id for the country (null for other entities).
- "Year" - Year from 1990 to 2019.
- "Internet_usage" -  The share of the entity's population who have used the internet in the last three months.

#### people
- "Entity" - The name of the country, region, or group.
- "Code" - Unique id for the country (null for other entities).
- "Year" - Year from 1990 to 2020.
- "Users" - The number of people who have used the internet in the last three months for that country, region, or group.

#### broadband
- "Entity" - The name of the country, region, or group.
- "Code" - Unique id for the country (null for other entities).
- "Year" - Year from 1998 to 2020.
- "Broadband_Subscriptions" - The number of fixed subscriptions to high-speed internet at downstream speeds >= 256 kbit/s for that country, region, or group.

_**Acknowledgments**: Max Roser, Hannah Ritchie, and Esteban Ortiz-Ospina (2015) - "Internet." OurWorldInData.org._

## Principal's questions

1. What are the top 5 countries with the highest internet use (by population share)?
2. How many people had internet access in those countries in 2019?
3. What are the top 5 countries with the highest internet use for each of the following regions: 'Africa Eastern and Southern', 'Africa Western and Central', 'Latin America & Caribbean', 'East Asia & Pacific', 'South Asia', 'North America', 'European Union'?
4. Create a visualization for those five regions' internet usage over time.
5. What are the 5 countries with the most internet users?
6. What is the correlation between internet usage (population share) and broadband subscriptions for 2019?
7. Summarize your findings.

### Importing modules

In [1]:
import pandas as pd
import numpy as np
import plotly.graph_objs as go
import plotly.offline as po
po.init_notebook_mode(connected=True)

### Loading and assessing broadband table

In [2]:
# Read the broadband table
broadband = pd.read_csv('data/broadband.csv')

# table preview
broadband.head()

Entity Code  Year  Broadband_Subscriptions
0  Afghanistan  AFG  2004                 0.000809
1  Afghanistan  AFG  2005                 0.000858
2  Afghanistan  AFG  2006                 0.001892
3  Afghanistan  AFG  2007                 0.001845
4  Afghanistan  AFG  2008                 0.001804

In [3]:
# Checking for null values across variables
broadband.isnull().sum()

Entity                       0
Code                       271
Year                         0
Broadband_Subscriptions      0
dtype: int64

In [4]:
# Counting observations with zero (0) broadband subscription
(broadband.Broadband_Subscriptions == 0).sum()

10

In [5]:
# Preview observations with no broadband subscription
broadband[broadband.Broadband_Subscriptions == 0]

Entity Code  Year  Broadband_Subscriptions
651                           Chad  TCD  2020                      0.0
856   Democratic Republic of Congo  COD  2010                      0.0
1454                         Haiti  HTI  2012                      0.0
1455                         Haiti  HTI  2013                      0.0
1456                         Haiti  HTI  2014                      0.0
1902                       Lebanon  LBN  2002                      0.0
1903                       Lebanon  LBN  2003                      0.0
1904                       Lebanon  LBN  2004                      0.0
1905                       Lebanon  LBN  2005                      0.0
1906                       Lebanon  LBN  2006                      0.0

In [6]:
# Counting unique broadband entity
len(broadband.Entity.unique())

222

In [7]:
# Counting unique broadband code
len(broadband.Code.unique())

209

In [8]:
# Counting unique broadband year
len(broadband.Year.unique())

23

In [9]:
# Checking general info including data type
broadband.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3888 entries, 0 to 3887
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Entity                   3888 non-null   object 
 1   Code                     3617 non-null   object 
 2   Year                     3888 non-null   int64  
 3   Broadband_Subscriptions  3888 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 121.6+ KB


### Loading and assessing internet table

In [10]:
# Read the internet table
internet = pd.read_csv('data/internet.csv')

# Table preview
internet.head()

Entity Code  Year  Internet_Usage
0  Afghanistan  AFG  1990             0.0
1  Afghanistan  AFG  1991             0.0
2  Afghanistan  AFG  1992             0.0
3  Afghanistan  AFG  1993             0.0
4  Afghanistan  AFG  1994             0.0

In [11]:
# Checking for null values across variables
internet.isnull().sum()

Entity               0
Code              1328
Year                 0
Internet_Usage       0
dtype: int64

In [12]:
# Counting observations with zero (0) internet usage
(internet['Internet_Usage'] == 0).sum()

994

In [13]:
# View of year with count of no internet usage
internet[internet['Internet_Usage'] == 0]['Year'].value_counts()

1990    226
1991    195
1992    178
1993    160
1994    128
1995     81
1996      5
2004      2
2003      2
2002      2
1997      2
2000      1
2001      1
1999      1
1998      1
2005      1
2006      1
2007      1
2008      1
2009      1
2010      1
2011      1
2012      1
2013      1
Name: Year, dtype: int64

In [14]:
# Previewing first 20 entity and the count of no internet usage
internet[internet['Internet_Usage'] == 0]['Entity'].value_counts()[:20]

North Korea              23
Timor                     9
Syria                     7
Sudan                     7
Eritrea                   7
Oman                      7
Comoros                   7
Gabon                     7
Guinea-Bissau             6
Papua New Guinea          6
Paraguay                  6
Rwanda                    6
Saint Kitts and Nevis     6
Samoa                     6
Haiti                     6
Guyana                    6
Gibraltar                 6
Sao Tome and Principe     6
Grenada                   6
Palestine                 6
Name: Entity, dtype: int64

In [15]:
# Counting unique internet entity
len(internet.Entity.unique())

261

In [16]:
# Counting unique internet code
len(internet.Code.unique())

215

In [17]:
# Counting unique internet year
len(internet.Year.unique())

30

In [18]:
# Checking general info including data type
internet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7089 entries, 0 to 7088
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Entity          7089 non-null   object 
 1   Code            5761 non-null   object 
 2   Year            7089 non-null   int64  
 3   Internet_Usage  7089 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 221.7+ KB


### Loading and assessing people table

In [19]:
# Read the people table
people = pd.read_csv('data/people.csv')

# Table preview
people.head()

Entity Code  Year  Users
0  Afghanistan  AFG  1990      0
1  Afghanistan  AFG  1991      0
2  Afghanistan  AFG  1992      0
3  Afghanistan  AFG  1993      0
4  Afghanistan  AFG  1994      0

In [20]:
# Checking for null values across variables
people.isnull().sum()

Entity      0
Code      307
Year        0
Users       0
dtype: int64

In [21]:
# Counting observations with where no one used the internet
(people.Users == 0).sum()

922

In [22]:
# View of year with count of no user of the internet
people[people.Users == 0]['Year'].value_counts()

1990    193
1991    178
1992    165
1993    153
1994    127
1995     80
1996      5
2004      2
2003      2
2002      2
1997      2
2000      1
2001      1
1999      1
1998      1
2005      1
2006      1
2007      1
2008      1
2009      1
2010      1
2011      1
2012      1
2013      1
Name: Year, dtype: int64

In [23]:
# View of year with count of no user of internet
people[people.Users == 0]['Entity'].value_counts()[:20]

North Korea         23
Timor                9
Oman                 7
Gabon                7
Eritrea              7
Comoros              7
Sudan                7
Syria                7
Madagascar           6
Malawi               6
Afghanistan          6
Maldives             6
Liechtenstein        6
Mali                 6
Marshall Islands     6
Mauritania           6
Mauritius            6
Lithuania            6
Latvia               6
Libya                6
Name: Entity, dtype: int64

In [24]:
# Counting unique people entity
len(people.Entity.unique())

223

In [25]:
# Counting unique people code
len(people.Code.unique())

214

In [26]:
# Counting unique people year
len(people.Year.unique())

31

In [27]:
# Checking general info and data type
people.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6379 entries, 0 to 6378
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Entity  6379 non-null   object
 1   Code    6072 non-null   object
 2   Year    6379 non-null   int64 
 3   Users   6379 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 199.5+ KB


## Data issues

### Tidiness issues
`broadband table`
* variable headers are capitalize

`internet table`
* variable headers are capitalize

`people table`
* variable headers are capitalize

## Cleaning data
### Making copy of original data

In [28]:
# copies of original data
broad_clean = broadband.copy()
int_clean = internet.copy()
ppl_clean = people.copy()

### Tidiness
Issue #1: Variable headers are capitalize in the broadband table

Issue #2: Variable headers are capitalize in the internet table

Issue #3: Variable headers are capitalize in the people table

**Define**: Variable headers should be in lower case for easy access. Change variables to lower case.

### Code

In [29]:
# Changing column headers for each table
broad_clean.columns = broad_clean.columns.str.lower()
int_clean.columns = int_clean.columns.str.lower()
ppl_clean.columns = ppl_clean.columns.str.lower()

### Test

In [30]:
# Print variables for each table
print(broad_clean.columns)
print(int_clean.columns)
print(ppl_clean.columns)

Index(['entity', 'code', 'year', 'broadband_subscriptions'], dtype='object')
Index(['entity', 'code', 'year', 'internet_usage'], dtype='object')
Index(['entity', 'code', 'year', 'users'], dtype='object')


### Merging Data
While assessing the tables above, the unique count for Entity, Code and Year was different for each table. Because of this, I will use an `Outer Join` to merge the table in order to keep all observations.   

In [31]:
# Merge all table
all_table = pd.merge(broad_clean, int_clean, how='outer', on=['entity', 'code', 'year'])\
    .merge(ppl_clean, how='outer', on=['entity', 'code', 'year'])
    
# Merge preview
all_table.head()

entity code  year  broadband_subscriptions  internet_usage     users
0  Afghanistan  AFG  2004                 0.000809        0.105809   24922.0
1  Afghanistan  AFG  2005                 0.000858        1.224148  298829.0
2  Afghanistan  AFG  2006                 0.001892        2.107124  536114.0
3  Afghanistan  AFG  2007                 0.001845        1.900000  492163.0
4  Afghanistan  AFG  2008                 0.001804        1.840000  486261.0

### Saving Data

In [32]:
# Save new table as a csv file
all_table.to_csv('all_table.csv')

### Answering Principal's Questions

#### 1. What are the top 5 countries with the highest internet use (by population share)?

In [33]:
# Group by countrie and summed their internet usage
top5Country = all_table.groupby('entity', as_index=False)[['internet_usage']].sum().sort_values('internet_usage', ascending=False).reset_index(drop=True)[:5]

# Print result
print("The top 5 countries with the internet use by (by population share are): \n{}".format(list(top5Country.entity)))
top5Country

The top 5 countries with the internet use by (by population share are): 
['Norway', 'Iceland', 'Denmark', 'Netherlands', 'Sweden']


entity  internet_usage
0       Norway     1855.540181
1      Iceland     1806.512910
2      Denmark     1753.807290
3  Netherlands     1735.916862
4       Sweden     1718.799370

#### 2. How many people had internet access in those countries in 2019?

In [34]:
# list of top 5 country from the previous answer
country_list = [i for i in top5Country.entity.tolist()]

# Filtering the table for 2019 observations and countries in the list above 
top5Country_users_2019 = all_table[(all_table['year'] == 2019) & (all_table['entity'].isin(country_list))] 

# Adding all users to get the total for the top 5 countries in year 2019
total_users = int(top5Country_users_2019.users.sum())
print("The total number of people from the top 5 countries who had access to internet in the year 2019 is {:,}".format(total_users))

The total number of people from the top 5 countries who had access to internet in the year 2019 is 37,181,605


#### 3. What are the top 5 countries with the highest internet use for each of the following regions: 'Africa Eastern and Southern', 'Africa Western and Central', 'Latin America & Caribbean', 'East Asia & Pacific', 'South Asia', 'North America', 'European Union'?

In [35]:
# Grouping countries into their region
AfricaEastern_andSouthern = ['Angola', 'Botswana', 'Lesotho', 'Mozambique', 'Namibia', 'South Africa', 'Zambia', 'Zimbabwe', 
                             'Burundi', 'Comoros', 'Djibouti', 'Eritrea', 'Ethiopia', 'Kenya', 'Madagascar', 'Malawi', 
                             'Mauritius', 'Rwanda', 'Seychelles', 'Somalia', 'South Sudan', 'Eswatini', 'Tanzania', 'Uganda', 
                             'Sierra Leone']

AfricaWestern_andCentral =['Benin', 'Burkina Faso', 'Cape Verde', "Cote d'Ivoire", 'Gambia', 'Ghana', 'Guinea', 'Guinea-Bissau',
                           'Liberia', 'Mali', 'Mauritania', 'Niger', 'Nigeria', 'Senegal', 'Togo', 'Cameroon', 'Central African Republic', 
                           'Chad', 'Congo', 'Democratic Republic of Congo', 'Equatorial Guinea', 'Gabon', 'Sao Tome and Principe']

LatinAmerica_andCaribbean = ['Brazil', 'Mexico', 'Colombia', 'Argentina', 'Peru', 'Venezuela', 'Chile', 'Guatemala', 'Ecuador', 'Bolivia',	
                             'Haiti', 'Cuba', 'Dominican Republic', 'Honduras', 'Paraguay', 'Nicaragua', 'El Salvador', 'Costa Rica', 
                             'Panama', 'Uruguay', 'Guyana', 'Suriname']

EastAsia_andPacific = ['Australia', 'Brunei', 'Cambodia', 'China', 'Fiji', 'French Polynesia', 'Guam', 'Hong Kong', 'Indonesia', 
                       'Japan', 'Kiribati', 'Laos', 'Macao', 'Malaysia', 'Marshall Islands', 'Micronesia (country)', 'Mongolia', 'Myanmar', 
                       'Nauru', 'New Caledonia', 'New Zealand', 'Palau', 'Papua New Guinea', 'Philippines', 'Samoa', 'Singapore', 'Solomon Islands', 
                       'Thailand', 'Timor', 'Tonga', 'Tuvalu', 'Vanuatu', 'Vietnam']

SouthAsia = ['Afghanistan', 'Bangladesh', 'Bhutan', 'India', 'Maldives', 'Nepal', 'Pakistan', 'Sri Lanka']

NorthAmerica = ['Antigua and Barbuda', 'Aruba', 'Bahamas', 'Barbados', 'Belize', 'Bermuda', 'British Virgin Islands', 'Canada', 'Cayman Islands', 
                'Costa Rica', 'Cuba', 'Dominica', 'Dominican Republic', 'El Salvador', 'Greenland', 'Georgia', 'Grenada', 'Guatemala', 'Haiti', 'Honduras', 
                'Jamaica', 'Mexico', 'Montserrat', 'Nicaragua', 'Panama', 'Puerto Rico', 'Saint Kitts and Nevis', 'Saint Lucia', 
                'Saint Vincent and the Grenadines', 'Trinidad and Tobago', 'United States', 'United States Virgin Islands']

EuropeanUnion = ['Albania', 'Andorra', 'Austria', 'Belgium', 'Bulgaria', 'Croatia', 'Cyprus', 'Czechia', 'Denmark', 'Estonia', 'Faeroe Islands', 'Finland', 'France', 
                 'Germany', 'Greece', 'Hungary', 'Ireland', 'Italy', 'Latvia', 'Lithuania', 'Luxembourg', 'Malta', 'Netherlands', 'Poland', 
                 'Portugal', 'Romania', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Bosnia and Herzegovina', 'Curacao', 'Gibraltar', 'Iceland', 'Liechtenstein', 
                 'Norway'] 

In [36]:
# conditions to identify countries in the table for each region
conds = [all_table['entity'].isin(AfricaEastern_andSouthern), all_table['entity'].isin(AfricaWestern_andCentral), all_table['entity'].isin(LatinAmerica_andCaribbean),
         all_table['entity'].isin(EastAsia_andPacific), all_table['entity'].isin(SouthAsia), all_table['entity'].isin(NorthAmerica), all_table['entity'].isin(EuropeanUnion)]

# Region names for each condition
choices = ['Africa Eastern and Southern', 'Africa Western and Central', 'Latin America & Caribbean', 'East Asia & Pacific', 'South Asia', 'North America', 'European Union']

# Creating a new variable with region name for each country and "Other region" for regions not in the above grouping
all_table['region'] = np.select(conds, choices, 'Other Region')

# Table preview
all_table.sample(7)

entity code  year  broadband_subscriptions  internet_usage  \
5245        Greece  GRC  1999                      NaN        6.877292   
1248         Gabon  GAB  2012                 0.285766       24.000000   
7150       Somalia  SOM  1994                      NaN        0.000000   
3573  Turkmenistan  TKM  2013                 0.033542        9.600000   
1813        Kuwait  KWT  2011                 1.483558       65.769071   
1415     Guatemala  GTM  2017                 3.139231       40.703049   
3737    Uzbekistan  UZB  2003                 0.010707        1.912595   

          users                       region  
5245   756080.0               European Union  
1248   440811.0   Africa Western and Central  
7150        0.0  Africa Eastern and Southern  
3573   533770.0                 Other Region  
1813  2067666.0                 Other Region  
1415  6283334.0    Latin America & Caribbean  
3737   495475.0                 Other Region

In [37]:
# Group table by region and country, then sum their internet usage
top5 = all_table.groupby(['region', 'entity'], as_index=False)['internet_usage'].sum()
# Sort the grouped data in descending order
top5 = top5.sort_values(['region', 'internet_usage'], ascending=False).groupby('entity').head()
# Preview
top5.head(10)

region          entity  internet_usage
271    South Asia        Maldives      477.745956
269    South Asia          Bhutan      214.550912
274    South Asia       Sri Lanka      184.654700
270    South Asia           India      162.838798
273    South Asia        Pakistan      159.785249
272    South Asia           Nepal      124.399907
268    South Asia      Bangladesh       75.535925
267    South Asia     Afghanistan       57.886490
247  Other Region     South Korea     1622.852927
261  Other Region  United Kingdom     1573.242646

In [38]:
# Filtering Africa Eastern and Southern Region
top5_AfricaEastern_andSouthern = top5[top5['region'] == 'Africa Eastern and Southern'].reset_index(drop=True)[:5]
print("In Africa Eastern and Southern Region, the top 5 countries with the highest internet use are; {}".format(list(top5_AfricaEastern_andSouthern.entity)))
top5_AfricaEastern_andSouthern

In Africa Eastern and Southern Region, the top 5 countries with the highest internet use are; ['Seychelles', 'Mauritius', 'South Africa', 'Botswana', 'Eswatini']


region        entity  internet_usage
0  Africa Eastern and Southern    Seychelles      623.893292
1  Africa Eastern and Southern     Mauritius      620.430432
2  Africa Eastern and Southern  South Africa      442.571872
3  Africa Eastern and Southern      Botswana      259.690147
4  Africa Eastern and Southern      Eswatini      223.023392

In [39]:
# Filtering Africa Western and Central Region
top5_AfricaWestern_andCentral = top5[top5['region'] == 'Africa Western and Central'].reset_index(drop=True)[:5]
# Print result
print("In Africa Western and Central Region, the top 5 countries with the highest internet use are; {}".format(list(top5_AfricaWestern_andCentral.entity)))
top5_AfricaWestern_andCentral

In Africa Western and Central Region, the top 5 countries with the highest internet use are; ['Cape Verde', 'Gabon', 'Sao Tome and Principe', 'Ghana', 'Senegal']


region                 entity  internet_usage
0  Africa Western and Central             Cape Verde      400.378133
1  Africa Western and Central                  Gabon      311.524724
2  Africa Western and Central  Sao Tome and Principe      308.740663
3  Africa Western and Central                  Ghana      187.846697
4  Africa Western and Central                Senegal      182.928232

In [40]:
# Filtering Latin America and Caribbean Region
top5_LatinAmerica_andCaribbean = top5[top5['region'] == 'Latin America & Caribbean'].reset_index(drop=True)[:5]
# Print result
print("In Latin America and Caribbean Region, the top 5 countries with the highest internet use are; {}".format(list(top5_LatinAmerica_andCaribbean.entity)))
top5_LatinAmerica_andCaribbean

In Latin America and Caribbean Region, the top 5 countries with the highest internet use are; ['Uruguay', 'Chile', 'Costa Rica', 'Brazil', 'Panama']


region      entity  internet_usage
0  Latin America & Caribbean     Uruguay      877.245839
1  Latin America & Caribbean       Chile      813.912766
2  Latin America & Caribbean  Costa Rica      802.481438
3  Latin America & Caribbean      Brazil      704.496071
4  Latin America & Caribbean      Panama      677.545370

In [41]:
top5_EastAsia_andPacific = top5[top5['region'] == 'East Asia & Pacific'].reset_index(drop=True)[:5]
print("In East Asia & Pacific Region, the top 5 countries with the highest internet use are; {}".format(list(top5_EastAsia_andPacific.entity)))
top5_EastAsia_andPacific

In East Asia & Pacific Region, the top 5 countries with the highest internet use are; ['Singapore', 'Japan', 'New Zealand', 'Hong Kong', 'Australia']


region       entity  internet_usage
0  East Asia & Pacific    Singapore     1438.394600
1  East Asia & Pacific        Japan     1421.882340
2  East Asia & Pacific  New Zealand     1419.584627
3  East Asia & Pacific    Hong Kong     1411.394392
4  East Asia & Pacific    Australia     1205.061203

In [42]:
top5_SouthAsia = top5[top5['region'] == 'South Asia'].reset_index(drop=True)[:5]
print("In South Asia Region, the top 5 countries with the highest internet use are; {}".format(list(top5_SouthAsia.entity)))
top5_SouthAsia

In South Asia Region, the top 5 countries with the highest internet use are; ['Maldives', 'Bhutan', 'Sri Lanka', 'India', 'Pakistan']


region     entity  internet_usage
0  South Asia   Maldives      477.745956
1  South Asia     Bhutan      214.550912
2  South Asia  Sri Lanka      184.654700
3  South Asia      India      162.838798
4  South Asia   Pakistan      159.785249

In [43]:
top5_NorthAmerica = top5[top5['region'] == 'North America'].reset_index(drop=True)[:5]
print("In North America Region, the top 5 countries with the highest internet use are; {}".format(list(top5_NorthAmerica.entity)))
top5_NorthAmerica

In North America Region, the top 5 countries with the highest internet use are; ['Bermuda', 'Canada', 'United States', 'Greenland', 'Barbados']


region         entity  internet_usage
0  North America        Bermuda     1505.171500
1  North America         Canada     1463.081215
2  North America  United States     1454.713237
3  North America      Greenland     1101.891829
4  North America       Barbados     1018.043396

In [44]:
top5_EuropeanUnion = top5[top5['region'] == 'European Union'].reset_index(drop=True)[:5]
print("In European Union Region, the top 5 countries with the highest internet use are; {}".format(list(top5_EuropeanUnion.entity)))
top5_EuropeanUnion

In European Union Region, the top 5 countries with the highest internet use are; ['Norway', 'Iceland', 'Denmark', 'Netherlands', 'Sweden']


region       entity  internet_usage
0  European Union       Norway     1855.540181
1  European Union      Iceland     1806.512910
2  European Union      Denmark     1753.807290
3  European Union  Netherlands     1735.916862
4  European Union       Sweden     1718.799370

#### 4. Create a visualization for those five regions' internet usage over time.

In [96]:
# list of top 5 Africa Eastern and Southern Region countries with high internet use
lst1 = list(top5_AfricaEastern_andSouthern.entity)
# Filtering table for countries in the list above
five_group1 = all_table[all_table['entity'].isin(lst1)]
# pivot table with countries as variables
five_group1 = five_group1.pivot_table('internet_usage', 'year', 'entity')

# Plotting 
# Function to make a line plot 
def add_region(region, color):
    fig.add_trace(go.Scatter(x=five_group1.index, y=five_group1[region],
                             mode="lines+markers",
                             marker= dict(
                                 symbol="diamond", size=3
                             ),
                            name=region, connectgaps=True,
                            line=dict(color=color, width=1)))
# Use function to make line plot for each country
fig = go.Figure()
add_region('Botswana', 'firebrick')
add_region('Mauritius', 'royalblue')
add_region('Eswatini', 'goldenrod')
add_region('Seychelles', 'darkturquoise')
add_region('South Africa', 'forestgreen')
fig.update_layout(title='Internet Usage in African Eastern and Southern Region', width=850,
                  yaxis_title = 'Internet Usage',
                  yaxis = dict(
                      tickvals = [0, 10, 20, 30, 40, 50, 60],
                      ticktext = ['0%', '10%', '20%', '30%', '40%', '50%', '60%']
                  )
                  )

In [95]:
# list of top 5 Africa Western and Central Region countries with high internet use
lst2 = list(top5_AfricaWestern_andCentral.entity)
# Filtering table for countries in the list above
five_group2 = all_table[all_table['entity'].isin(lst2)]
# pivot table with countries as variables
five_group2 = five_group2.pivot_table('internet_usage', 'year', 'entity')

# Plotting 
# Function to make a line plot 
def add_region(region, color):
    fig.add_trace(go.Scatter(x=five_group2.index, y=five_group2[region],
                             mode="lines+markers",
                             marker= dict(
                                 symbol="diamond", size=3
                             ),
                            name=region, connectgaps=True, 
                            line=dict(color=color, width=1)))
# Use function to make line plot for each country
fig = go.Figure()
add_region('Cape Verde', 'firebrick')
add_region('Gabon', 'royalblue')
add_region('Ghana', 'goldenrod')
add_region('Sao Tome and Principe', 'darkturquoise')
add_region('Senegal', 'forestgreen')
fig.update_layout(title='Internet Usage in Africa Western and Central Region', width=850,
                  yaxis_title = 'Internet Usage',
                  yaxis = dict(
                      tickvals = [0, 10, 20, 30, 40, 50, 60],
                      ticktext = ['0%', '10%', '20%', '30%', '40%', '50%', '60%']
                  )
                  )

In [93]:
# list of top 5 Latin America and Caribbean Region countries with high internet use
lst3 = list(top5_LatinAmerica_andCaribbean.entity)
# Filtering table for countries in the list above
five_group3 = all_table[all_table['entity'].isin(lst3)]
# pivot table with countries as variables
five_group3 = five_group3.pivot_table('internet_usage', 'year', 'entity')

# Plotting 
# Function to make a line plot 
def add_region(region, color):
    fig.add_trace(go.Scatter(x=five_group3.index, y=five_group3[region],
                             mode="lines+markers",
                             marker= dict(
                                 symbol="diamond", size=3
                             ),
                            name=region, connectgaps=True, 
                            line=dict(color=color, width=1)))
# Use function to make line plot for each country
fig = go.Figure()
add_region('Brazil', 'firebrick')
add_region('Chile', 'royalblue')
add_region('Costa Rica', 'goldenrod')
add_region('Panama', 'darkturquoise')
add_region('Uruguay', 'forestgreen')
fig.update_layout(title='Internet Usage in Latin America and Caribbean Region', width=850,
                  yaxis_title = 'Internet Usage',
                  yaxis = dict(
                      tickvals = [0, 20, 40, 60, 80],
                      ticktext = ['0%', '20%', '40%', '60%', '80%']
                  )
                  )

In [92]:
# list of top 5 East Asia and Pacific Region countries with high internet use
lst4 = list(top5_EastAsia_andPacific.entity)
# Filtering table for countries in the list above
five_group4 = all_table[all_table['entity'].isin(lst4)]
# pivot table with countries as variables
five_group4 = five_group4.pivot_table('internet_usage', 'year', 'entity')

# Plotting 
# Function to make a line plot 
def add_region(region, color):
    fig.add_trace(go.Scatter(x=five_group4.index, y=five_group4[region],
                             mode="lines+markers",
                             marker= dict(
                                 symbol="diamond", size=3
                             ),
                            name=region, connectgaps=True,
                            line=dict(color=color, width=1)))
# Use function to make line plot for each country
fig = go.Figure()
add_region('Australia', 'firebrick')
add_region('Hong Kong', 'royalblue')
add_region('Japan', 'goldenrod')
add_region('New Zealand', 'darkturquoise')
add_region('Singapore', 'forestgreen')
fig.update_layout(title='Internet Usage in East Asia and Pacific Region', width=850,
                  yaxis_title = 'Internet Usage',
                  yaxis = dict(
                      tickvals = [0, 20, 40, 60, 80],
                      ticktext = ['0%', '20%', '40%', '60%', '80%']
                  )
                  )

In [90]:
# list of top 5 South Asia Region countries with high internet use
lst5 = list(top5_SouthAsia.entity)
# Filtering table for countries in the list above
five_group5 = all_table[all_table['entity'].isin(lst5)]
# pivot table with countries as variables
five_group5 = five_group5.pivot_table('internet_usage', 'year', 'entity')

# Plotting 
# Function to make a line plot
def add_region(region, color):
    fig.add_trace(go.Scatter(x=five_group5.index, y=five_group5[region],
                             mode="lines+markers",
                             marker= dict(
                                 symbol="diamond", size=3
                             ),
                            name=region, connectgaps=True,
                            line=dict(color=color, width=1)))
# Use function to make line plot for each country
fig = go.Figure()
add_region('Bhutan', 'firebrick')
add_region('India', 'royalblue')
add_region('Maldives', 'goldenrod')
add_region('Pakistan', 'darkturquoise')
add_region('Sri Lanka', 'forestgreen')
fig.update_layout(title='Internet Usage in South Asia Region', width=850,
                  yaxis_title = 'Internet Usage',
                  yaxis = dict(
                      tickvals = [0, 10, 20, 30, 40, 50, 60],
                      ticktext = ['0%', '10%', '20%', '30%', '40%', '50%', '60%']
                  )
                  )

In [86]:
# list of top 5 North America Region countries with high internet use
lst6 = list(top5_NorthAmerica.entity)
# Filtering table for countries in the list above
five_group6 = all_table[all_table['entity'].isin(lst6)]
# pivot table with countries as variables
five_group6 = five_group6.pivot_table('internet_usage', 'year', 'entity')

# Plotting 
# Function to make a line plot
def add_region(region, color):
    fig.add_trace(go.Scatter(x=five_group6.index, y=five_group6[region],
                             mode="lines+markers",
                             marker= dict(
                                 symbol="diamond", size=3
                             ),
                            name=region, connectgaps=True,
                            line=dict(color=color, width=1)))
# Use function to make line plot for each country
fig = go.Figure()
add_region('Barbados', 'firebrick')
add_region('Bermuda', 'royalblue')
add_region('Canada', 'goldenrod')
add_region('Greenland', 'darkturquoise')
add_region('United States', 'forestgreen')
fig.update_layout(title='Internet Usage in North America Region ', width=850,
                  yaxis_title = 'Internet Usage',
                  yaxis = dict(
                      tickvals = [0, 20, 40, 60, 80, 100],
                      ticktext = ['0%', '20%', '40%', '60%', '80%', '100%']
                  )
                  )

In [89]:
# list of top 5 European Union Region countries with high internet use
lst7 = list(top5_EuropeanUnion.entity)
# Filtering table for countries in the list above
five_group7 = all_table[all_table['entity'].isin(lst7)]
# pivot table with countries as variables
five_group7 = five_group7.pivot_table('internet_usage', 'year', 'entity')

# Plotting 
# Function to make a line plot
def add_region(region, color):
    fig.add_trace(go.Scatter(x=five_group7.index, y=five_group7[region], 
                             mode="lines+markers",
                             marker= dict(
                                 symbol="diamond", size=3
                             ),
                            name=region, connectgaps=True,
                            line=dict(color=color, width=1)))
# Use function to make line plot for each country
fig = go.Figure()
add_region('Denmark', 'firebrick')
add_region('Iceland', 'royalblue')
add_region('Netherlands', 'goldenrod')
add_region('Norway', 'darkturquoise')
add_region('Sweden', 'forestgreen')
fig.update_layout(title='Internet Usage in European Union Region', width=850,
                  yaxis_title = 'Internet Usage'
                  )

#### 5. What are the 5 countries with the most internet users?

In [124]:
# Filtering out continent and non countries
no_code = all_table[~((all_table['code'].isnull()) | (all_table['entity'] == 'World'))]
# Group by entity and sort in descending order
top5_CountriesMostUsers = no_code.groupby('entity', as_index=False)['users'].sum().sort_values('users', ascending=False).reset_index(drop=True)[:5]
# Print result
print("The 5 countries with most internet users are; {}".format(list(top5_CountriesMostUsers['entity'])))
print(top5_CountriesMostUsers)

# Plotting visual
fig = go.Figure()
fig.add_trace(go.Bar(x=top5_CountriesMostUsers.entity, y=top5_CountriesMostUsers.users, text=top5_CountriesMostUsers.users/1000000000, 
                     texttemplate="%{text:.3}B", marker=dict(color='firebrick')))
fig.update_layout(title='Countries with the most internet users', width=850,
                  yaxis_title = 'Users (Billions)')
fig.update_yaxes(showticklabels=False)

The 5 countries with most internet users are; ['China', 'United States', 'India', 'Japan', 'Brazil']
          entity         users
0          China  9.269901e+09
1  United States  5.070162e+09
2          India  2.942444e+09
3          Japan  2.041997e+09
4         Brazil  1.735975e+09


#### 6. What is the correlation between internet usage (population share) and broadband subscriptions for 2019?


In [54]:
# Filtering 2019 records
record_2019 = all_table[all_table['year'] == 2019]
# Finding correlation
correlation = record_2019['internet_usage'].corr(record_2019['broadband_subscriptions'])
# Print result
print("The correlation between internet usage (population share) and broadband subscriptions for 2019 is {}".format(correlation))

The correlation between internet usage (population share) and broadband subscriptions for 2019 is 0.5590077105170114


## Summary
1. The top 5 countries with the highest internet use (by population share) are; **Norway**, **Iceland**, **Denmark**, **Netherlands** and **Sweden**.
2. The number of people who had access to internet for the above countries in 2019 is **37,181,605**.
3. Below are the regions and their top 5 countries with the highest internet use

    | Africa Eastern and Southern Region|  Africa Western and Central Region | Latin America and Caribbean Region | East Asia & Pacific Region | South Asia Region| North America Region | European Union Region | 
    |:-----------------------:|:-----------------------:| :-----------------------:| :-----------------------:| :-----------------------:|:-----------------------:| :-----------------------:| 
    | Seychelles **=** 623.893292 | Cape Verde **=** 400.378133| Uruguay **=** 877.245839 | Singapore **=** 1438.394600 | Maldives **=** 477.745956 | Bermuda **=** 1505.171500 | Norway **=** 1855.540181 | 
    | Mauritius **=** 620.430432| Gabon **=** 311.524724 | Chile **=** 813.912766 | Japan **=** 1421.882340 | Bhutan **=** 214.550912 | Canada **=** 1463.081215 | Iceland **=** 1806.512910 | 
    | South Africa **=** 442.571872| Sao Tome and Principe **=** 308.740663| Costa Rica **=** 802.481438 | New Zealand **=** 1419.584627 | Sri Lanka **=** 184.654700 | United States **=** 1454.713237| Denmark **=** 1753.807290 | 
    | Botswana **=** 259.690147| Ghana **=** 187.846697 | Brazil **=** 704.496071 | Hong Kong **=** 1411.394392 | India **=** 162.838798| Greenland **=** 1101.891829 | Netherlands **=** 1735.916862 |
    | Eswatini **=** 223.023392 | Senegal **=** 182.928232 | Panama **=** 677.545370 | Australia **=** 1205.061203 | Pakistan **=** 159.785249 | Barbados **=** 1018.043396 | Sweden **=** 1718.799370 | 
>

4. Below is the summary of the top 5 countries latest internet usage in each region
* In African Eastern and Southern region, Seychelles and South Africa Uses below 50% of the internet while Botswana uses above 60%.
* In Africa Western and Central region, Ghana, Sao Tome and Principe, with Senegal uses below 40% of the internet. Cape Verde and Gabon uses above 50%.
* In the Latin America and Caribben region, Brazil, Panama and Uruaguay falls in  range of 60% to 80% of internet usage. Chile and Costa Rica uses above 80%
* In the East Asia, the top five countrie, Australia, Hong Kong, Japan, New Zealand and Singapore uses above 80% of the internet.
* In South Asia, amongst the top five countries, only Maldives uses above 60% of the internet. Bhutan, India, Pakistan and Sri Lanka uses blow 45% of the internet.
* In North America region, Bermuda and Canada uses above 90% of the internet. United States 88.4% and Barbados 82% and Greenland 69.4%
* In European Union, all five top countries uses above 90% of the internet.
5. The 5 countries with the most internet user are; 

    | Country| Users (Billions) |  
    |:-----------------------:|:-----------------------:|
    | China | 9.27 | 
    | United States | 5.07 |  
    | India | 2.94 |  
    | Japan| 2.04 |  
    | Brazil | 1.74 |  
>
6. The correlation between internet usage (population share) and broadband subcriptions for 2019 is 0.559.